[clang-format] Add `AllowShortNamespacesOnASingleLine` option

galenelias opened the pull request:
This fixes #101363 which is a resurrection of a previously opened but never completed review: https://reviews.llvm.org/D11851

The feature is to allow code like the following not to be broken across multiple lines:

```
namespace foo { class bar; }
namespace foo { namespace bar { class baz; } }
```

Code like this is commonly used for forward declarations, which are ideally kept compact. This is also apparently the format that include-what-you-use will insert for forward declarations.

Also, fix an off-by-one error in `CompactNamespaces` code. For nested namespaces with 3 or more namespaces, it was incorrectly compacting lines which were 1 or two spaces over the `ColumnLimit`, leading to incorrect formatting results.

Commit:
clang-format: Add "AllowShortNamespacesOnASingleLine" option

This addresses: https://github.com/llvm/llvm-project/issues/101363
which is a resurrection of a previously opened but never completed
review: https://reviews.llvm.org/D11851

The feature is to allow code like the following not to be broken across
multiple lines:

```
namespace foo { class bar; }
namespace foo { namespace bar { class baz; } }
```

Code like this is commonly used for forward declarations, which are
ideally kept compact. This is also apparently the format that
include-what-you-use will insert for forward declarations.

mydeveloperday commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -616,6 +625,62 @@ class LineJoiner {
+      if (I[1]->Last->is(TT_LineComment))
+        return 0;
+
+      unsigned inner_limit = Limit - I[1]->Last->TotalLength - 3;
```
Local variables are normally `PascalCase`

Commit:
Use PascalCase for local variable names

HazardyKnusperkeks commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -616,6 +625,62 @@ class LineJoiner {
+      if (!MergedLines)
+        return 0;
+      // check if there is even a line after the inner result
+      if (I + 2 + MergedLines >= E)
```
Iterators are normally not allowed to move past end. How about something like `std::distance(I, E) <= MergedLines + 2`?

HazardyKnusperkeks commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -916,6 +981,23 @@ class LineJoiner {
+  bool nextNLinesFitInto(SmallVectorImpl<AnnotatedLine *>::const_iterator I,
+                         SmallVectorImpl<AnnotatedLine *>::const_iterator E,
+                         unsigned Limit) {
+    unsigned joinedLength = 0;
```
```suggestion
    unsigned JoinedLength = 0;
```

HazardyKnusperkeks commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -916,6 +981,23 @@ class LineJoiner {
+                         SmallVectorImpl<AnnotatedLine *>::const_iterator E,
+                         unsigned Limit) {
+    unsigned joinedLength = 0;
+    for (SmallVectorImpl<AnnotatedLine *>::const_iterator J = I + 1; J != E;
```
```suggestion
    for (auto J = I + 1; J != E;
```

Commit:
Address review comments

mydeveloperday commented on clang/include/clang/Format/Format.h:
```diff
@@ -972,6 +972,11 @@ struct FormatStyle {
 
+  /// If ``true``, ``namespace a { class b; }`` can be put on a single a single
+  /// line.
+  /// \version 19
```
19 is now wrong it should be 20

Commit:
Switch version of new format option to 20

owenca reviewed (commented):
The following are missing:
- Run `dump_format_style.py`.
- Add a `ConfigParseTest` for the new option.

How does the new option interact with `CompactNamespaces`? For example:
- `AllowShortNamespacesOnASingleLine: true` and `CompactNamespaces: false`
```
namespace a {
namespace b { class c; }
} // namespace a
```
- Both `true`
`namespace a { namespace b { class c; } }`

galenelias commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -27981,6 +27981,118 @@ TEST_F(FormatTest, BreakBinaryOperations) {
+  // No ColumnLimit, allows long nested one-liners, but also leaves multi-line
+  // instances alone.
+  Style.ColumnLimit = 0;
+  verifyFormat("namespace foo { namespace bar { namespace baz { namespace qux "
+               "{ class quux; } } } }",
```
The clang-formatting rules seem to preclude this.  It will always pull the code into what I have here, which is having the string split across two lines and starting on the line with the verifyFormat.

Commit:
Re-generate StylClangFormatStyleOptions.rst

Commit:
Add ConfigParseTest

Commit:
Fix interaction with CompactNamespaces

galenelias commented:
> The following are missing:
> 
> * Run `dump_format_style.py`.
> * Add a `ConfigParseTest` for the new option.
> 
> How does the new option interact with `CompactNamespaces`? For example:
> 
> * `AllowShortNamespacesOnASingleLine: true` and `CompactNamespaces: false`
> 
> ```
> namespace a {
> namespace b { class c; }
> } // namespace a
> ```
> 
> * Both `true`
>   `namespace a { namespace b { class c; } }`

Aah, I hadn't considered the interaction of this AllowShortNamespacesOnASingleLine and CompactNamespaces, as AllowShortNamespacesOnASingleLine effectively forces CompactNamespaces behavior when the block can all fit on a single line.  However, it does appear that they were in conflict where the CompactNamespaces code was running first and then short circuiting out of the AllowShortNamespacesOnASingleLine logic.

Moved the logic around a bit, as AllowShortNamespacesOnASingleLine is mostly a superset of the CompactNamespaces logic.  There are still some ambiguous situations around nested namespaces with a single statement inside that could be merged with both these two rules.  I am just letting those get merged with CompactNamespaces first if the full expression can't be pulled onto a single line.

owenca commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -616,6 +626,62 @@ class LineJoiner {
+      const unsigned InnerLimit = Limit - I[1]->Last->TotalLength - 3;
+      const unsigned MergedLines = tryMergeNamespace(I + 1, E, InnerLimit);
+      if (!MergedLines)
+        return 0;
```
```suggestion
        return 0;
        const auto N = MergedLines + 2;
```

owenca commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -616,6 +626,62 @@ class LineJoiner {
+      if (!MergedLines)
+        return 0;
+      // Check if there is even a line after the inner result.
+      if (std::distance(I, E) <= MergedLines + 2)
```
```suggestion
      if (std::distance(I, E) <= N)
```

owenca commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -616,6 +626,62 @@ class LineJoiner {
+        return 0;
+      // Check that the line after the inner result starts with a closing brace
+      // which we are permitted to merge into one line.
+      if (I[2 + MergedLines]->First->is(tok::r_brace) &&
+          !I[2 + MergedLines]->First->MustBreakBefore &&
+          !I[1 + MergedLines]->Last->is(TT_LineComment) &&
+          nextNLinesFitInto(I, I + 2 + MergedLines + 1, Limit)) {
+        return 2 + MergedLines;
```
```suggestion
      if (I[N]->First->is(tok::r_brace) &&
          !I[N]->First->MustBreakBefore &&
          I[MergedLines + 1]->Last->isNot(tok::comment) &&
          nextNLinesFitInto(I, I + N + 1, Limit)) {
        return N;
```

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -27981,6 +27981,132 @@ TEST_F(FormatTest, BreakBinaryOperations) {
+      "namespace foo { namespace bar { namespace baz { class qux; } } }",
+      Style);
+
+  verifyNoChange("namespace foo {\n"
+                 "namespace bar { namespace baz { class qux; } }\n"
+                 "}",
+                 Style);
```
```suggestion
// FIXME
#if 0
  verifyNoChange("namespace foo {\n"
                 "namespace bar { namespace baz { class qux; } }\n"
                 "}",
                 Style);
#endif
```
Shouldn't this be merged as well? See this potential [bug](https://discourse.llvm.org/t/clang-format-compactnamespaces-in-combination-with-columnlimit-0/78808).

galenelias replied:
Well, you would probably be more the expert on the intended design of 'ColumnLimit: 0', but my mental model of ColumnLimit: 0 is basically that it will respect the line breaks in the existing code as long as they fall on valid 'seams'.  This lets you make really long lines, but also lets you produce very short lines by the same token, because it's basically taking the existing wrapping as the truth, assuming it's a valid wrapping.  I personally really value this behavior in the code-base I work on, so I wouldn't consider these instances a bug (same goes for the linked discourse link).

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -27981,6 +27981,132 @@ TEST_F(FormatTest, BreakBinaryOperations) {
+  Style = BaseStyle;
+  Style.FixNamespaceComments = true;
+
+  verifyFormat(
+      "namespace foo {\n"
+      "namespace bar { namespace baz { class qux; } } // namespace bar\n"
+      "} // namespace foo",
+      "namespace foo { namespace bar { namespace baz { class qux; } } }",
+      Style);
+
+  // This option doesn't really make any sense with ShortNamespaceLines = 0.
+  Style.ShortNamespaceLines = 0;
+
+  verifyFormat(
+      "namespace foo {\n"
+      "namespace bar {\n"
+      "namespace baz { class qux; } // namespace baz\n"
+      "} // namespace bar\n"
+      "} // namespace foo",
+      "namespace foo { namespace bar { namespace baz { class qux; } } }",
+      Style);
+}
```
IMO it makes more sense that `AllowShortNamespacesOnASingleLine` disables `FixNamespaceComments` on the merged closing braces.

galenelias replied:
While I agree on principal, I think that interaction is currently beyond my knowledge of how to achieve within the design of clang-format.  The code that does the FixNamespaceComments happens before the UnwrappedLineFormatter, so you would need to somehow to call 'tryMergeNamespace' from the `NamespaceEndCommentsFixer` code to see if the namespace will be merged, then avoid adding the comment there.

I think it's OK to leave this interaction sub-optimal for now, with the guidance that people should use the 'namespace Foo::Bar' syntax with this feature if they want to avoid poor interactions with FixNamespaceComments.

If this is more straightforward than I am understanding, I'm happy to try to make a reasonable change to accommodate it.

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -27981,6 +27981,132 @@ TEST_F(FormatTest, BreakBinaryOperations) {
+
+  // If we can't merge an outer nested namespaces, but can merge an inner
+  // nested namespace, then CompactNamespaces will merge the outer namespace
+  // first, preventing the merging of the inner namespace
```
```suggestion
  // first, preventing the merging of the inner namespace.
```

galenelias commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -616,6 +626,62 @@ class LineJoiner {
+    // Check if it's a namespace inside a namespace, and call recursively if so
+    // '3' is the sizes of the whitespace and closing brace for " _inner_ }".
+    if (I[1]->First->is(tok::kw_namespace)) {
+      if (I[1]->Last->is(TT_LineComment))
```
Yah, fair question.  I think I could see it either way.  I'm fine starting with not supporting trailing block comments either and then could relax it if there are explicit requests for that behavior.

galenelias commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -616,6 +626,62 @@ class LineJoiner {
 
+  unsigned tryMergeNamespace(SmallVectorImpl<AnnotatedLine *>::const_iterator I,
+                             SmallVectorImpl<AnnotatedLine *>::const_iterator E,
+                             unsigned Limit) {
+    if (Limit == 0)
```
We can't take an `int` here, because when ColumnLimit = 0, then the calling code passes in `UINT_MAX` which would then underflow the int.  I could technically take a int64_t, but in general it seems like the style is to pre-check for underflow before doing subtraction, which is the approach here (although it's a little indirect).  See below.

galenelias commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -616,6 +626,62 @@ class LineJoiner {
+      if (I[1]->Last->is(TT_LineComment))
+        return 0;
+
+      const unsigned InnerLimit = Limit - I[1]->Last->TotalLength - 3;
+      const unsigned MergedLines = tryMergeNamespace(I + 1, E, InnerLimit);
+      if (!MergedLines)
```
This underflow is prevented by prior call to `nextTwoLinesFitInto`, which will ensure that Limit is large enough such that this won't underflow.

Commit:
Address additional PR comments

Commit:
clang-format run

galenelias commented:
Ping

owenca commented:
> > The following are missing:
> > 
> > * Run `dump_format_style.py`.
> > * Add a `ConfigParseTest` for the new option.
> > 
> > How does the new option interact with `CompactNamespaces`? For example:
> > 
> > * `AllowShortNamespacesOnASingleLine: true` and `CompactNamespaces: false`
> > 
> > ```
> > namespace a {
> > namespace b { class c; }
> > } // namespace a
> > ```
> > 
> > * Both `true`
> >   `namespace a { namespace b { class c; } }`
> 
> Aah, I hadn't considered the interaction of this AllowShortNamespacesOnASingleLine and CompactNamespaces, as AllowShortNamespacesOnASingleLine effectively forces CompactNamespaces behavior when the block can all fit on a single line. However, it does appear that they were in conflict where the CompactNamespaces code was running first and then short circuiting out of the AllowShortNamespacesOnASingleLine logic.
> 
> Moved the logic around a bit, as AllowShortNamespacesOnASingleLine is mostly a superset of the CompactNamespaces logic. There are still some ambiguous situations around nested namespaces with a single statement inside that could be merged with both these two rules. I am just letting those get merged with CompactNamespaces first if the full expression can't be pulled onto a single line.

IMO `AllowShortNamespacesOnASingleLine` shouldn't imply `CompactNamespaces` so that we can get any of the formats below:

- `AllowShortNamespacesOnASingleLine: false`, `CompactNamespaces: false`
```
namespace a {
namespace b {
class c;
}
}
```

- `AllowShortNamespacesOnASingleLine: false`, `CompactNamespaces: true`
```
namespace a { namespace b {
class c;
}}
```

- `AllowShortNamespacesOnASingleLine: true`, `CompactNamespaces: false`
```
namespace a {
namespace b { class c; }
}
```

- `AllowShortNamespacesOnASingleLine: true`, `CompactNamespaces: true`
```
namespace a { namespace b { class c; } }
```

owenca reviewed (commented):
Please update the release notes.

owenca commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -616,6 +627,62 @@ class LineJoiner {
+    // Check if it's a namespace inside a namespace, and call recursively if so.
+    // '3' is the sizes of the whitespace and closing brace for " _inner_ }".
+    if (I[1]->First->is(tok::kw_namespace)) {
+      if (I[1]->Last->is(tok::comment))
```
```suggestion
      if (I[1]->Last->is(tok::comment) || !Style.CompactNamespaces)
```
And adjust the test cases and/or turning `CompactNamespaces` on/off. See https://github.com/llvm/llvm-project/pull/105597#issuecomment-2408847951.

Commit:
Re-work interaction with 'CompactNamespaces' option

- Disable recursive merging of nested namespaces onto a single line
  unless 'CompactNamespaces' is enabled.

- Fix off-by-one error in CompactNamespaces code. For nested namespaces
  with 3 or more namespaces, it was incorrectly compacting lines which
  were 1 or two spaces over the ColumnLimit, leading to incorrect
  formatting results.

- Add release notes mention.

- Re-generate Style Options documentation

galenelias commented:
> I'm inclined to think that all the AllowShort options should work the same way whether ColumnLimit is 0. However, a couple of them I checked (AllowShortLoops and AllowShortBlocks) have no effect unless set to false. I suggest that we mark the test as a FIXME or don't include it here.

As an active user of ColumnLimit 0, I disagree.  At least for my team we use ColumnLimit of 0 to allow for more customization and variability of the column limit.  Treating it like ColumnLimit: Infinity causes huge constructs to be pulled into a single line for some of these 'AllowShort' options, which we actively don't want.  Making somewhat arbitrary and contextual decisions about how to do the breaking is a 'feature' in my opinion - although I'm not sure if that's truly the intent and vision behind ColumnLimit 0.

My thought here is that this feature should function similarly to how CompactNamespaces interacts with ColumnLimit: 0, which is basically that it allows arbitrary line breaking.  So you could have a 1000 character line, or you could break it after every namespace.  Forcing it to always pull single statements no matter how long onto a single line would be undesirable in my opinion.

I'm definitely open to changing things here, and being consistent if there is some other vision of how features are supposed to interact with ColumnLimit 0, but this has been my mental model so far, and that seems to be the natural behavior that I've observed, so validating that functionality in the tests seems reasonable to me.

owenca commented:
> > I'm inclined to think that all the AllowShort options should work the same way whether ColumnLimit is 0. However, a couple of them I checked (AllowShortLoops and AllowShortBlocks) have no effect unless set to false. I suggest that we mark the test as a FIXME or don't include it here.
> 
> As an active user of ColumnLimit 0, I disagree. At least for my team we use ColumnLimit of 0 to allow for more customization and variability of the column limit. Treating it like ColumnLimit: Infinity causes huge constructs to be pulled into a single line for some of these 'AllowShort' options, which we actively don't want. Making somewhat arbitrary and contextual decisions about how to do the breaking is a 'feature' in my opinion - although I'm not sure if that's truly the intent and vision behind ColumnLimit 0.

Below is the documentation on [`ColumnLimit`](https://releases.llvm.org/19.1.0/tools/clang/docs/ClangFormatStyleOptions.html#columnlimit):
> A column limit of `0` means that there is no column limit. In this case, clang-format will respect the input’s line breaking decisions within statements unless they contradict other rules.

The term _statements_ above can be loosely interpreted as _unwrapped lines_. Since all the `AllowShort` options involve merging _wrapped_ ines, they should not be ignored even when `ColumnLimit` is set to 0. (See also [`
RemoveEmptyLinesInUnwrappedLines `](https://clang.llvm.org/docs/ClangFormatStyleOptions.html#removeemptylinesinunwrappedlines).)

> My thought here is that this feature should function similarly to how CompactNamespaces interacts with ColumnLimit: 0, which is basically that it allows arbitrary line breaking. So you could have a 1000 character line, or you could break it after every namespace. Forcing it to always pull single statements no matter how long onto a single line would be undesirable in my opinion.

IMO a better way to have arbitrary line breakings is to add a `Leave` suboption.

> I'm definitely open to changing things here, and being consistent if there is some other vision of how features are supposed to interact with ColumnLimit 0, but this has been my mental model so far, and that seems to be the natural behavior that I've observed, so validating that functionality in the tests seems reasonable to me.

I don't think we want to add any tests whose formats are questionable or still under discussion. WDYT @mydeveloperday @HazardyKnusperkeks @rymiel ?

Commit:
Remove ColumnLimit=0 tests

Commit:
Remove blank line in release notes

galenelias commented:
Ping

owenca commented on clang/include/clang/Format/Format.h:
```diff
@@ -988,6 +988,10 @@ struct FormatStyle {
 
+  /// If ``true``, ``namespace a { class b; }`` can be put on a single line.
+  /// \version 20
+  bool AllowShortNamespacesOnASingleLine;
```
You need to add it to `operator==()`.

owenca reviewed (commented):
Please rebase and run `ninja clang-format-style`.

galenelias commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -616,6 +627,63 @@ class LineJoiner {
+      assert(Limit >= I[1]->Last->TotalLength + 3);
+      const unsigned InnerLimit = Limit - I[1]->Last->TotalLength - 3;
+      const unsigned MergedLines = tryMergeNamespace(I + 1, E, InnerLimit);
+      if (!MergedLines)
```
Sorry, missed the `if (MergedLines == 0)` suggestion.  The diffing UI on GitHub PRs is surprisingly blunt.

Commit:
PR review comments

Commit:
Clang-format previous commit

owenca commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -361,9 +361,19 @@ class LineJoiner {
     // FIXME: There are probably cases where we should use FirstNonComment
     // instead of TheLine->First.
 
+    if (TheLine->Last->is(tok::l_brace)) {
+      if (Style.AllowShortNamespacesOnASingleLine &&
+          TheLine->First->is(tok::kw_namespace)) {
+        unsigned result = tryMergeNamespace(I, E, Limit);
+        if (result > 0)
+          return result;
+      }
+    }
```
```suggestion
    if (Style.AllowShortNamespacesOnASingleLine &&
        TheLine->First->is(tok::kw_namespace) &&
        TheLine->Last->is(tok::l_brace)) {
      const auto result = tryMergeNamespace(I, E, Limit);
      if (result > 0)
        return result;
    }
```

Commit:
PR comment cleanup

owenca reviewed (commented):
Also, additional cleanup of `tryMergeNamespace()`:
```cpp
     if (Limit == 0)
       return 0;
-    if (I[1]->InPPDirective != (*I)->InPPDirective ||
-        (I[1]->InPPDirective && I[1]->First->HasUnescapedNewline)) {
+
+    assert(I[1]);
+    const auto &L1 = *I[1];
+    if (L1.InPPDirective != (*I)->InPPDirective ||
+        (L1.InPPDirective && L1.First->HasUnescapedNewline)) {
       return 0;
     }
-    if (I + 2 == E || I[2]->Type == LT_Invalid)
+
+    if (std::distance(I, E) <= 2)
+      return 0;
+
+    assert(I[2]);
+    const auto &L2 = *I[2];
+    if (L2.Type == LT_Invalid)
       return 0;
 
     Limit = limitConsideringMacros(I + 1, E, Limit);
@@ -645,13 +654,13 @@ private:
 
     // Check if it's a namespace inside a namespace, and call recursively if so.
     // '3' is the sizes of the whitespace and closing brace for " _inner_ }".
-    if (I[1]->First->is(tok::kw_namespace)) {
-      if (I[1]->Last->is(tok::comment) || !Style.CompactNamespaces)
+    if (L1.First->is(tok::kw_namespace)) {
+      if (L1.Last->is(tok::comment) || !Style.CompactNamespaces)
         return 0;
 
-      assert(Limit >= I[1]->Last->TotalLength + 3);
-      const unsigned InnerLimit = Limit - I[1]->Last->TotalLength - 3;
-      const unsigned MergedLines = tryMergeNamespace(I + 1, E, InnerLimit);
+      assert(Limit >= L1.Last->TotalLength + 3);
+      const auto InnerLimit = Limit - L1.Last->TotalLength - 3;
+      const auto MergedLines = tryMergeNamespace(I + 1, E, InnerLimit);
       if (MergedLines == 0)
         return 0;
       const auto N = MergedLines + 2;
@@ -672,11 +681,11 @@ private:
     // line.
 
     // The line which is in the namespace should end with semicolon.
-    if (I[1]->Last->isNot(tok::semi))
+    if (L1.Last->isNot(tok::semi))
       return 0;
 
     // Last, check that the third line starts with a closing brace.
-    if (I[2]->First->isNot(tok::r_brace) || I[2]->First->MustBreakBefore)
+    if (L2.First->isNot(tok::r_brace) || L2.First->MustBreakBefore)
       return 0;
```
Use more descriptive names instead of `L1` and `L2` if you like.

galenelias commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -4504,6 +4504,16 @@ TEST_F(FormatTest, FormatsCompactNamespaces) {
                "} // namespace aaaaaa",
                Style);
+
+  verifyFormat("namespace a { namespace b { namespace c {\n"
+               "}}} // namespace a::b::c",
+               Style);
+
+  verifyFormat("namespace a { namespace b {\n"
+               "namespace cc {\n"
+               "}}} // namespace a::b::cc",
```
I'll defer to your judgement, but I personally think it is helpful to test each side of these edge cases around line wrapping.  Having a test to verify things don't get wrapped at 'X' columns, but then having tests cases which are like X/2 long doesn't really validate the precision of the width calculations, as evidenced by this bug, and means we might be wrapping too aggressively, and won't catch that via any existing tests.  If we only verify lines significantly shorter or longer than the ColumnLimit then we it is easy for mistakes in the logic to sneak through.

owenca replied:
I agree, but in this particular case the bug occurs when the length is over the column limit, not when it's within the limit?

The following test case for `ColumnLimit: 41`:
```
namespace a { namespace b { namespace cc {
}}} // namespace a::b::cc
```
is equivalent to the test case below with `ColumnLimit: 40`:
```
namespace a { namespace b { namespace c {
}}} // namespace a::b::c
```

If you want to test that your fix doesn't break the existing behavior when the length is at the limit, I suggest that you reuse the existing test case:
```cpp
   verifyFormat("namespace out {\n"
                "int i;\n"
                "namespace in {\n"
                "int j;\n"
                "} // namespace in\n"
                "int k;\n"
                "} // namespace out",
                "namespace out { int i;\n"
                "namespace in { int j; } // namespace in\n"
                "int k; } // namespace out",
                Style);
 
+  Style.ColumnLimit = 41;
   verifyFormat("namespace A { namespace B { namespace C {\n"
                "}}} // namespace A::B::C",
                "namespace A { namespace B {\n"
                "namespace C {\n"
                "}} // namespace B::C\n"
                "} // namespace A",
                Style);
 
   Style.ColumnLimit = 40;
   verifyFormat("namespace aaaaaaaaaa {\n"
                "namespace bbbbbbbbbb {\n"
                "}} // namespace aaaaaaaaaa::bbbbbbbbbb",
                "namespace aaaaaaaaaa {\n"
                "namespace bbbbbbbbbb {\n"
                "} // namespace bbbbbbbbbb\n"
                "} // namespace aaaaaaaaaa",
                Style);
```
FormatTest.cpp is 28K+ lines long and still growing. It should be broken up into multiple files. Until then, I prefer that we don't add nice-to-have new tests that aren't absolutely related to the pull request.

Commit:
More PR comment cleanup

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -28314,6 +28320,116 @@ TEST_F(FormatTest, KeepFormFeed) {
+  // concatenated namespace syntax instead. e.g. 'namespace foo::bar'
+  Style = BaseStyle;
+  Style.FixNamespaceComments = true;
+
```
```suggestion
```

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -28314,6 +28320,116 @@ TEST_F(FormatTest, KeepFormFeed) {
+  verifyFormat("namespace foo { namespace bar { class baz; } }", Style);
+
+  // Without CompactNamespaces, we won't merge consecutive namespace
+  // declarations
```
```suggestion
  // declarations.
```

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -28314,6 +28320,116 @@ TEST_F(FormatTest, KeepFormFeed) {
+  // too long to fit within the ColumnLimit, reducing the how likely the line
+  // will still fit on a single line. The recommendation for now is to use the
+  // concatenated namespace syntax instead. e.g. 'namespace foo::bar'
+  Style = BaseStyle;
```
```suggestion
```

owenca commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -28314,6 +28320,116 @@ TEST_F(FormatTest, KeepFormFeed) {
 }
 
+TEST_F(FormatTest, ShortNamespacesOption) {
+  auto BaseStyle = getLLVMStyle();
+  BaseStyle.AllowShortNamespacesOnASingleLine = true;
+  BaseStyle.FixNamespaceComments = false;
+  BaseStyle.CompactNamespaces = true;
+
+  auto Style = BaseStyle;
```
```suggestion
  auto Style = getLLVMStyle();
  Style.AllowShortNamespacesOnASingleLine = true;
  Style.CompactNamespaces = true;
  Style.FixNamespaceComments = false;
```

owenca commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -616,6 +626,71 @@ class LineJoiner {
+                             unsigned Limit) {
+    if (Limit == 0)
+      return 0;
+    assert(I[1]);
```
```suggestion

    assert(I[1]);
```

owenca reviewed (commented):
Final NFC cleanup.

owenca committed:
Apply suggestions from code review

Final NFC cleanup

owenca committed:
Update FormatTest.cpp
